src/validation.cpp

```diff
@@ -3925,6 +3930,28 @@ static bool ContextualCheckBlockHeader(const CBlockHeader& block, BlockValidatio
     return true;
 }
 
+/**
+ * We want to enforce certain rules (specifically the 64-byte transaction check)
+ * before we call CheckBlock to check the merkle root. This allows us to enforce
+ * malleability checks which may interact with other CheckBlock checks.
+ * This is currently called both in AcceptBlock prior to writing the block to
+ * disk and in ConnectBlock.
+ * Note that as this is called before merkle-tree checks so must never return a
+ * non-malleable error condition.
+ */
+static bool ContextualBlockPreCheck(const CBlock& block, BlockValidationState& state, const ChainstateManager& chainman, const CBlockIndex* pindexPrev)
```

Comment:
Why is this complexity of performing contextual checks before `CheckBlock` necessary? I don't see a reason why we would need to not persist blocks invalid because they contain 64 bytes transactions as opposed to any other reason for invalidity.

Reply:
Here's a sketch of two alternative approaches:

 * https://github.com/ajtowns/bitcoin/commit/0d565eb0bd810f7526e6e6c6ffef5d9fe59186b3
   * adds contextual information to `CheckBlock()` (ie a chainman and pindexPrev pointer)
   * when all txs are 64 bytes, downgrade all errors to BLOCK_MUTATED
   * otherwise, if a 64 byte tx is found, and the 64 byte rule is enabled, gives a consensus failure
   * this would allow dropping the extra call to `CheckBlock()` in `ProcessNewBlock()`
 * https://github.com/ajtowns/bitcoin/commit/b601f934451872216a70eeac5592c19bd7f6b7f8
   * moves the check into `ContextualCheckBlock`
   * (requires keeping the extra call to `CheckBlock()` in `ProcessNewBlock()` in order to downgrade those failures to as-if they were BLOCK_MUTATED)